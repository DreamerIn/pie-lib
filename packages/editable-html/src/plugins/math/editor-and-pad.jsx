import { HorizontalKeypad } from '@pie-lib/math-input';

import React from 'react';
import debug from 'debug';
import PropTypes from 'prop-types';
import MathQuillEditor from './mathquill/editor';
import { withStyles } from '@material-ui/core/styles';

const log = debug('@pie-lib:editable-html:plugins:math:editor-and-pad');

const toNodeData = data => {
  if (!data) {
    return;
  }

  const { type, value } = data;

  if (type === 'command' || type === 'cursor') {
    return data;
  } else if (value === 'clear') {
    return { type: 'clear' };
  } else {
    return { type: 'write', value };
  }
};

export class EditorAndPad extends React.Component {
  static propTypes = {
    latex: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired,
    classes: PropTypes.object
  };

  onClick = data => {
    const c = toNodeData(data);
    log('mathChange: ', c);

    if (c.type === 'clear') {
      log('call clear...');
      this.input.clear();
    } else if (c.type === 'command') {
      this.input.command(c.value);
    } else if (c.type === 'cursor') {
      this.input.keystroke(c.value);
    } else {
      this.input.write(c.value);
    }
  };

  onEditorChange = latex => {
    const { onChange } = this.props;
    onChange(latex);
  };

  /** Only render if the mathquill instance's latex is different */
  shouldComponentUpdate(nextProps) {
    const inputIsDifferent = this.input.latex() !== nextProps.latex;
    log('[shouldComponentUpdate] ', 'inputIsDifferent: ', inputIsDifferent);
    return inputIsDifferent;
  }

  render() {
    const { latex, classes } = this.props;

    log('[render]', latex);

    return (
      <div className={classes.mathToolbar}>
        <MathQuillEditor
          className={classes.mathEditor}
          ref={r => (this.input = r)}
          latex={latex}
          onChange={this.onEditorChange}
        />
        <hr className={classes.hr} />
        <HorizontalKeypad onClick={this.onClick} />
      </div>
    );
  }
}

const styles = theme => ({
  mathEditor: {
    marginTop: theme.spacing.unit,
    marginBottom: theme.spacing.unit
  },
  hr: {
    padding: 0,
    margin: 0,
    height: '1px',
    border: 'none',
    borderBottom: `solid 1px ${theme.palette.primary.main}`
  },
  mathToolbar: {
    textAlign: 'center',
    '& > .mq-math-mode': {
      border: 'solid 0px lightgrey'
    },
    '& > .mq-focused': {
      outline: 'none',
      boxShadow: 'none',
      border: `dotted 1px ${theme.palette.primary.main}`,
      borderRadius: '0px'
    }
  }
});

export default withStyles(styles)(EditorAndPad);
