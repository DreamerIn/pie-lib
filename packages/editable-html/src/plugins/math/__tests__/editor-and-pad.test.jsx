import { EditorAndPad } from '../editor-and-pad';
import { shallow } from 'enzyme';
import React from 'react';

jest.mock('@pie-lib/math-input', () => ({
  HorizontalKeypad: () => <div>HorizontalKeypad</div>
}));

jest.mock('../mathquill/editor', () => () => <div>MockEditor</div>);

describe('EditorAndPad', () => {
  let onChange;

  beforeEach(() => {
    onChange = jest.fn();
  });

  const mkWrapper = extras => {
    const props = {
      latex: '\\frac{1}{2}',
      onChange,
      classes: {},
      ...extras
    };

    const w = shallow(<EditorAndPad {...props} />);
    w.instance().input = {
      clear: jest.fn(),
      command: jest.fn(),
      keystroke: jest.fn(),
      write: jest.fn()
    };
    return w;
  };

  describe('snapshot', () => {
    it('renders', () => {
      const w = mkWrapper();

      expect(w).toMatchSnapshot();
    });
  });

  describe('onClick', () => {
    let w;

    beforeEach(() => {
      w = mkWrapper();
    });

    it('calls clear', () => {
      const i = w.instance();
      i.onClick({ value: 'clear' });
      expect(i.input.clear).toHaveBeenCalled();
    });

    it('calls command', () => {
      const i = w.instance();
      i.onClick({ type: 'command', value: 'foo' });
      expect(i.input.command).toHaveBeenCalledWith('foo');
    });

    it('calls keystroke', () => {
      const i = w.instance();
      i.onClick({ type: 'cursor', value: '<' });
      expect(i.input.keystroke).toHaveBeenCalledWith('<');
    });

    it('calls keystroke', () => {
      const i = w.instance();
      i.onClick({ type: 'other', value: 'write data' });
      expect(i.input.write).toHaveBeenCalledWith('write data');
    });
  });
});
