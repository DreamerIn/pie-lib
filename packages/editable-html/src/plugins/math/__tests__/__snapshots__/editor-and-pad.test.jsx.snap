// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorAndPad snapshot renders 1`] = `
<div>
  <Component
    latex="\\\\frac{1}{2}"
    onChange={[Function]}
  />
  <hr />
  <HorizontalKeypad
    onClick={[Function]}
  />
</div>
`;
